:memo: Change Docksal doc into tutorial (#2259)
platformsh/platformsh-docs: docs/src/development/local/docksal.md

---
title: Use Docksal for local development
sidebarTitle: Docksal
weight: 4
---

[Docksal](https://docksal.io) is a Docker-based local development tool that plays nicely with Platform.sh.
It allows you to have fully containerized environments to run everything locally
without having to install tools (such as the Platform.sh CLI) on your machine.
It's maintained by a community of developers and is a viable option for most Platform.sh projects.

This guide assumes you have a project already running within Platform.sh.

> If you're starting from scratch, first [create a project from a template](https://console.platform.sh/projects/create-project/template).

Author: That'd be done in: https://github.com/platformsh/platformsh-docs/issues/2260

For more information on `fin`, run `fin help`.

## 2. Create a projects directory

Create a directory to store all of your Docksal projects.
It's recommended to be `~/Projects` on Linux/macOS and `C:\Projects` on Windows.
See all [restrictions on the projects directory](https://docs.docksal.io/getting-started/project-setup/#projects-directory).

## 3. Add an API token

To connect Docksal with your Platform.sh account, use a Platform.sh API token.

1. [Create an API token](./../cli/api-tokens.md#get-a-token) in the console.
2. Add the token to your Docksal configuration by running this command:

   ```bash
   fin config set --global SECRET_PLATFORMSH_CLI_TOKEN="<API_TOKEN>"
   ```

Now you can run `fin platform <command>` from the your computer without needing to install the Platform.sh CLI.

## 4. Get your project

To get your project, pull it with the following command:

```bash
fin pull init --hosting-platform='platformsh' --hosting-site=<PROJECT_ID> --hosting-env=<ENVIRONMENT_NAME> <TARGET_DIRECTORY_NAME>
```

This creates a directory with the specified name with all your files and code.
It also adds a `.docksal` directory with all necessary Docksal configuration.

> These files are ignored by Platform.sh.

Reviewer: ❓ Are they added to the `.gitgnore` automatically as well? or should it be manually added, just as the [docksal-local.yml](https://docs.docksal.io/stack/understanding-stack-config/#docksal-local)?

Author: I don't think you'd want to ignore them in Git. That way various people can test things locally under the same conditions. `docksal-local` files are explicitly different.

## 5. Add commands

Docksal doesn't automatically copy over any commands you have in your [build flavor](../../administration/../configuration/app/app-reference.md#build)
and [hooks](../../configuration/app/hooks/_index.md).

> To get you project running like on Platform.sh, you have to add the commands to Docksal.

Reviewer: ⛏️ Second repetition of "like" in line 63, maybe adding a few more words sounds nicer?
```suggestion
To get your project running in the same way as on Platform.sh, you have to add the commands to Docksal.
```

Author: I think having a conjunction like "like" a couple times in separate paragraphs isn't so bad, but the second one seems easier to change without making it as long as "in the same way as".

Reviewer: 😍 The last change is much better than my proposal.

You can add new commands here that run on `fin init`
or create a new file to run on `fin <NAME_OF_FILE>`.

You can add your build flavor and hook commands after these commands.
That way the Docksal CLI container is running and can process commands.

For example, to install dependencies, use a command like the following depending on your dependency manager:

{{< codetabs >}}

---
title=npm
file=none
highlight=false
---

```bash {location=".docksal/commands/init"}
#!/usr/bin/env bash

fin project rm -f
fin project start

cd "$PROJECT_ROOT"
echo "Installing dependencies..."
npm install

```
<--->

---
title=Composer
file=none
highlight=false
---

```bash {location=".docksal/commands/init"}
#!/usr/bin/env bash

fin project rm -f
fin project start

cd "$PROJECT_ROOT"
echo "Installing dependencies..."
fin composer install

```

{{< /codetabs >}}

## 6. Run your project

Now your project is ready to run.
In your project directory, run `fin init`.

The command returns the project URL `http://<DIRECTORY_NAME>.docksal/`.
Open this URL to see your project running.

## What's next

You've got a project running on a local web server.

> Now you can add customizations.

Reviewer: 💅 An example could be to add `mkcert` to have https on docksal.
https://docs.docksal.io/tools/mkcert/#body-inner

Author: The examples were intended to be the following sections.

For more ideas and options, see the [Docksal documentation](https://docs.docksal.io/).

### Add environment variables

If your project requires environment variables for its hooks,
add them to the Docksal environment.

{{< codetabs >}}

---
title=Using the CLI
file=none
highlight=bash
---

> fin config set <VARIBLE_NAME>=<VARIABLE_VALUE>

Reviewer: 📋

The surrounding lines could also be removed?
```suggestion
fin config set <VARIABLE_NAME>=<VARIABLE_VALUE>
```

<--->

---
title=Directly in the configuration file
file=none
highlight=false
---

```txt {location=".docksal/docksal.env"}

> <VARIBLE_NAME>=<VARIABLE_VALUE>

Reviewer: ```suggestion
<VARIABLE_NAME>=<VARIABLE_VALUE>
```

> 

Reviewer: 📋
```suggestion
```

HOSTING_PLATFORM="platformsh"
```

{{< /codetabs >}}

To apply your changes, run the following command:

```bash
fin project restart
```

### Customize your stack

Docksal comes configured with a default PHP container (known as the CLI container),
web container, and database container.
You can customize the versions for each.

Check the available versions for each by running `fin image registry <IMAGE>`.
For example, to check the available containers for the app, run `fin image registry docksal/cli`.

You can then use the version you want by adding it as an [environment variable](#add-environment-variables).

> For example, to use the latest version, run the following command:

Reviewer: 💅
A bullet list of examples might be nicer to read?

Author: I'm not sure I understand. There's one example, so it doesn't seem to fit well into a bullet list. There are two steps, but one example for each.

```bash
fin config set CLI_IMAGE='docksal/cli:latest'
```

You can also customize the `.docksal/docksal.yml` file.
Use it for a docker-compose definition for your [custom configurations](https://docs.docksal.io/stack/custom-configuration/#custom-configuration).

### Import MySQL data from Platform.sh into Docksal

To download your data from Platform.sh and load it into your Docksal database container, run the following commands:

```bash
fin platform db:dump --gzip -f /tmp/database.sql.gz

> fin exec 'zcat < /tmp/database.sql.gz | mysql -u user -puser -h db default'

Reviewer: ❓Might be worth to add variables there
```suggestion
fin exec 'zcat < /tmp/database.sql.gz | mysql -u <USERNAME> -p <PASSWORD> -h db <DATABASE_NAME>'
```
and point to [where the values can be retrieved](https://docs.docksal.io/service/db/access/).

Author: The link is to where they can be set rather than retrieved, but could help. It might be a little more advanced than I'd want to fit in a tutorial, especially a side note at the bottom. How about a link to where they could be set as a compromise?

Reviewer: 😍  I like the additional context you added, it should make it clearer how to adapt and use the command when needed.
